### PR TITLE
Complete rewrite of the MCS algorithm

### PR DESCRIPTION
**Pull request by proteneer**
This PR completely rewrites our MCS algorithm from scratch. There's no meaningful changes from the `bgl` repo, except for the addition of more tests.

### CODE REVIEW

**Review by maxentile — APPROVED** [review]

File: timemachine/fe/rbfe.py [hunk 1]
```diff
@@ -738,7 +749,7 @@ def run_edge_and_save_results(
         return path
 
     path = f"success_rbfe_result_{edge.mol_a_name}_{edge.mol_b_name}.pkl"
-    pkl_obj = (mol_a, mol_b, edge.metadata, smarts, core, solvent_res, solvent_top, complex_res, complex_top)
```
Comment (proteneer) [inline timemachine/fe/rbfe.py:741]: @jkausrelay sanity check: does this affect anything on the consumer end?

Comment (jkausrelay) [reply timemachine/fe/rbfe.py:741]: This would only affect @mcwitt , we call the run_* directly

**Comment by proteneer** [comment]
@jkausrelay  I implemented
1) complete_rings_only as a post-processing step (which requires a subsequent filtering of disconnected components to be sensible)
2) an additional deduplication check, since the usage of filters sometimes results in redundant cores

can you 👁️ or test this out once more?

File: timemachine/fe/atom_mapping.py [hunk 2]
```diff
@@ -1,290 +1,435 @@
...
+
+
+def _uniquify_core(core):
+    core_list = []
+    for a, b in core:
+        core_list.append((a, b))
+    return frozenset(core_list)
+
+
+def _deduplicate_all_cores(all_cores):
```
Comment (proteneer) [inline timemachine/fe/atom_mapping.py:277]: @jkausrelay dedup logic

File: timemachine/fe/atom_mapping.py [hunk 3]
```diff
@@ -1,290 +1,435 @@
...
+    mcg_a = mcgregor.Graph(mol_a.GetNumAtoms(), get_edges(mol_a))
+    mcg_b = mcgregor.Graph(mol_b.GetNumAtoms(), get_edges(mol_b))
+
+    for a, b in core:
+        # subgraph and graph are consistent
+        pred_sgg_a = a_cycles[a] == sg_a_cycles[a]
+        pred_sgg_b = b_cycles[b] == sg_b_cycles[b]
+        # a and b are consistent
+        pred_sg_ab = sg_a_cycles[a] == sg_b_cycles[b]
+        pred_g_ab = a_cycles[a] == b_cycles[b]
```
Comment (proteneer) [inline timemachine/fe/atom_mapping.py:230]: @jkausrelay invariance for complete rings:

We require a strong consistency that each mol's subgraph and graph have the same ring requirement, and that mol_a's subgraph is consistent with mol_b's subgraph. I think the last assertion that mol_a's graph is consistent with mol_b's graph may be redundant (and would be enforced by atom-atom predicates, which turns out to be unnecessary)

File: tests/test_atom_mapping.py [hunk 4]
```diff
@@ -1,67 +1,426 @@
...
+        ]
+    )
+
+    # 0-indexed
+    expected_core -= 1
+
+    all_cores_fzset = get_all_cores_fzset(all_cores)
+    assert tuples_to_set(expected_core) in all_cores_fzset
+
+    # TBD: if we have check_chiral == True then we should expect length of all_cores == 1
```
Comment (proteneer) [inline tests/test_atom_mapping.py:425]: @maxentile added a placeholder for you to ensure that if we check for chirality then only a single core-mapping should be valid.

Comment (maxentile) [reply tests/test_atom_mapping.py:425]: Thanks -- will take a closer look after the holiday!

(As of yesterday morning, was intending to re-enable the chiral check disabled by this PR in a subsequent PR using this branch: https://github.com/proteneer/timemachine/compare/bgl...chiral_check . Will also update any affected tests.)

File: tests/test_align.py [hunk 5]
```diff
@@ -1,240 +0,0 @@
...
-import pytest
-from rdkit import Chem
-from rdkit.Chem import AllChem
-
-from timemachine.constants import DEFAULT_FF
-from timemachine.fe.atom_mapping import get_core_by_mcs, get_core_with_alignment, mcs
-from timemachine.fe.topology import AtomMappingError
-from timemachine.fe.utils import set_romol_conf
-from timemachine.ff import Forcefield
-from timemachine.md import align
```
Comment (maxentile) [inline tests/test_align.py:13]: Is deleting this test file intentional?

(Wasn't deleted in initial version I initially reviewed, and some code it tests e.g. `md/align.py` was not removed by this PR)

Comment (proteneer) [reply tests/test_align.py:13]: I'm not sure how to make this compatible with the alignment code yet - I'll talk to @badisa and @jkausrelay later this week to decide how to deal with this correctly (it had a bunch of tests that I wasn't sure how to fix)

File: timemachine/fe/atom_mapping.py (hunk 3 above)
Comment (jkausrelay) [reply timemachine/fe/atom_mapping.py:234]: Makes sense to me

**Comment by jkausrelay** [comment]
> @jkausrelay I implemented
> 
> 1. complete_rings_only as a post-processing step (which requires a subsequent filtering of disconnected components to be sensible)
> 2. an additional deduplication check, since the usage of filters sometimes results in redundant cores
> 
> can you 👁️ or test this out once more?

Will test out the merged version with complete rings only

**Pull request merged**